browser/ui/views/search_engines/active_window_search_provider_manager.h

```diff
@@ -0,0 +1,60 @@
+/* Copyright (c) 2021 The Brave Authors. All rights reserved.
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+#ifndef BRAVE_BROWSER_UI_VIEWS_SEARCH_ENGINES_ACTIVE_WINDOW_SEARCH_PROVIDER_MANAGER_H_
+#define BRAVE_BROWSER_UI_VIEWS_SEARCH_ENGINES_ACTIVE_WINDOW_SEARCH_PROVIDER_MANAGER_H_
+
+#include "base/gtest_prod_util.h"
+#include "base/scoped_observation.h"
+#include "components/prefs/pref_member.h"
+#include "ui/views/widget/widget.h"
+#include "ui/views/widget/widget_observer.h"
+
+class Browser;
+class Profile;
+
+// Setting proper default search provider for profile of activated window.
+// Previously, we've used separate TemplateURLService for using different
+// provider for normal and private profile because we want to use different
+// provider for them. However, this caused many issues due to difference from
+// chromium's usage. Chromium uses one instance for both profile.
+// So, we will follow chromium's configuration(use one service instance for all
+// profile). Instead, we will set appropriate provider for currently active
+// window(profile - normal, private or tor). With this approach, we don't need
+// to use separate TemplareURLService for normal, private(tor) profile.
+// This doesn't handle guest window's provider because guest window only uses
+// its own private profile. Using SearchEngineProviderService for managing guest
+// window's DDG toggle button config is sufficient.
+class ActiveWindowSearchProviderManager : public views::WidgetObserver {
```

Comment:
I'm not quire sure this is a way to go. The service is queried for `GetDefaultSearchProvider()` across the code and the callers can get different results depending on a visible widget, isn't it strange?

Reply:
With this approach, we can use one service for normal/private like chromium.
Many places & services with single instance service (use one service instance for both profile) refers `TemplateURLService`. This makes conflict with our needs(we need different `TemplateURLService` instance).
Search provider extension crash also comes from using different `TemplateURLService` instance.
Of course, I also think this approach is not the best but this works well and we can archieve our needs w/o patching/overrding anymore.

Reply:
IMO, we should introduce new complex patches/overriding to fix crash issues(https://github.com/brave/brave-browser/issues/15224, https://github.com/brave/brave-browser/issues/14218) w/o this approach.
and I think this approach is straightforward and can maintain easily. WDYT? @bridiver

Reply:
I think I agree with @iefremov here. In addition to what @iefremov said, changing the result based on the active window could easily return the wrong result if there are any async calls to it. It seems very risky to me based on that.

Reply:
in fact I found instances where `GetDefaultSearchProvider` is called async which means the active window could change from what it was when the action was initiated

Reply:
will close this PR and I'll try to resolve with another approach - https://github.com/brave/brave-core/pull/8626